### DISCUSSION TITLE
Generating a virtual ECU with standardized subservice identifiers

### DISCUSSION THREAD

**Discussion by rumpelsepp** [Fraunhofer-AISEC/gallia#455] [Q&A] [answered]
@LeventCelik asked in https://github.com/Fraunhofer-AISEC/gallia/issues/453#issuecomment-1778413497

> Could you also help me with generating a virtual ECU with standardized subservice identifiers? For instance, using 0x10, SBF of 0x1 for default session, 0x3 for extended, etc. Can I do that, and if I can, which command shall I use?
>
> Secondly, are the services with 'null' (such as
"0x01 (ShowCurrentData)": null,
"0x02 (ShowFreezeFrameData)": null,
"0x03 (ShowstoredDiagnosticTroubleCodes)": null,
) not available?
>
> I am trying to understand the tool, but could not find these in the documentation.

/cc @fkglr

**Comment by fkglr** [comment, marked as answer]
>Could you also help me with generating a virtual ECU with standardized subservice identifiers? For instance, using 0x10, SBF of 0x1 for default session, 0x3 for extended, etc. Can I do that, and if I can, which command shall I use?

You can control some of the randomness of the ECU using its command line parameters. E.g. you can create a virtual ECU which has at least the three session 0x01, 0x02 and 0x03 using the parameter `--mandatory_sessions [1,2,3]`. You can also suppress creating further sessions by setting the probability of session activation to zero with `--p_session 0`.
These sessions however don't have a specific behavior. You would have to program that in on your own.
What you can do is to also specify available services for all sessions explicitly like so `--mandatory_services "[ReadDataByIdentifier, DiagnosticSessionControl, TesterPresent, EcuReset, ClearDiagnosticInformation]"`, or set the probability to one in order to activate all UDS services on each session `--p_service 1`.
For some of these services, there already is a meaningful implementation, however for some of them, there currently is not, so by default a GeneralReject UDS error will be returned.
Just in case you are intereseted, you could add support for further services in [server.py](https://github.com/Fraunhofer-AISEC/gallia/blob/master/src/gallia/services/uds/server.py#L361) for the class RandomUDSServer.

> Secondly, are the services with 'null' (such as
"0x01 (ShowCurrentData)": null,
"0x02 (ShowFreezeFrameData)": null,
"0x03 (ShowstoredDiagnosticTroubleCodes)": null,
) not available?

The lists behind the services refer to the available sub-functions of that service. Services which by specification do not have subfunctions have null listed instead.
Here you can also see which sessions are reachable out of a certain session. E.g. if for session 0x01 the map contains `"0x10 (DiagnosticSessionControl)": "['0x01', '0x02']"`, then you can switch to session 0x01 and 0x02 from this session 0x01 using the service DiagnosticSessionControl (0x10).

I hope this helps

**Reply by LeventCelik** [reply]
Thanks, that adds a lot of clarity! The current state of the documentation seem to go over these a bit fast, but now I understand.

Is there any public information about which services are fully implemented, or should I get that by probing over the available services and eliminating the ones that return general reject?
Are all the implemented services located at https://github.com/Fraunhofer-AISEC/gallia/blob/603cef0c263cf91af98e5e8318d1d8265767d300/src/gallia/services/uds/core/service.py ?

**Reply by rumpelsepp** [reply]
These services are relevant for the **client**. @fkglr correct me if I am wrong; the implemented services for the **server** are here: https://github.com/Fraunhofer-AISEC/gallia/blob/603cef0c263cf91af98e5e8318d1d8265767d300/src/gallia/services/uds/server.py

**Reply by fkglr** [reply]
That is correct. More precisely starting from https://github.com/Fraunhofer-AISEC/gallia/blob/603cef0c263cf91af98e5e8318d1d8265767d300/src/gallia/services/uds/server.py#L550.
